### ISSUE TITLE
Embedding uglymol in a Jupyter notebook

### ISSUE THREAD

**Issue by JBGreisman** [uglymol/uglymol#8]
I very frequently work with reflection data in Python using [reciprocalspaceship](https://github.com/Hekstra-Lab/reciprocalspaceship) and [gemmi](https://github.com/project-gemmi/gemmi), and one of my dreams would be to have a way to inspect the corresponding map interactively in a Jupyter notebook -- to make a small change to coefficients/phases and to then be able to view the changes in `uglymol` without needing to write an MTZ and open it up in visualization suite. 

Here is the sort of thing that I wish I could do: [notebook example](https://hekstra-lab.github.io/reciprocalspaceship/examples/6_timeresolvedPYP.html#Visualize-Weighted-Difference-Map). In reality, in that example I have pre-computed weighted/unweighted difference maps that are hosted elsewhere, and I just embed a page running uglymol in an `iframe`. However, if this occurred without the intermediate maps that are hosted elsewhere, one could imagine trying out different weighting schemes or scanning parameter values and observing the impacts on the map interactively.

I'm posting this to get your thoughts on whether this sort of thing seems feasible and whether you have any ideas for how to make it work. In my mind the core requirement to get this working would be passing a PDB/MTZ/map to uglymol without it being read from a file hosted elsewhere. Is there a way to do that? Or to read from a local file?

**Comment by wojdyr** [comment]
It'd possible to read a local file if there was a local server (I suppose Python's http.server would do) that would pass this file to the web browser. In principle, it'd also be possible to do it without writing a file if your program that calculates the maps was directly serving them to the web browser.
I don't have experience with setting up such things. Would you like to use it together with a jupyter notebook?

**Comment by JBGreisman** [comment]
You're correct that it is possible to serve local files in a fairly easy way using Python to pass files to the browser for viewing in `uglymol`. This [StackOverflow answer](https://stackoverflow.com/a/21957017) includes an example of how to do this for a standard Python http.server (with CORS headers). This can also be done in a jupyter notebook by starting the notebook with the argument `--NotebookApp.allow_origin='*'`.

My goal would be to use it in a jupyter notebook, and the latter option here does enable me to produce a working solution that involves writing/reading temporary files. It would be ideal to be able to do such a thing without needing to write a temporary file, but that may be more effort than its worth.

**Comment by wojdyr** [comment]
If it'd help I could add a function in gemmi that writes Mtz or Ccp4 map to a memory buffer and returns it as Python bytes or string.
But then it'd need to be passed somehow by the server to the browser as a file. I don't know if it'd be possible. And I don't know if it'd make a significant difference comparing with writing temporary files.

**Comment by JBGreisman** [comment]
thanks -- I'll try to implement this the way that I think I understand with temporary files, and will check back if its too slow to be feasible.
